tests/pushbits/main.c

```diff
@@ -76,6 +79,90 @@ echo()
 	exit(1);
 }
 
+static int
+wait_ms(size_t msec)
+{
+	struct timespec ts;
+
+	ts.tv_sec = msec / 1000;
+	ts.tv_nsec = (msec % 1000) * 1000000;
+	nanosleep(&ts, NULL);
+
+	return (0);
+}
+
+static int
+chain()
+{
+	pthread_t thread;
+	int in[2];
+	int out[2];
+	const char * msg = "ping";
+	size_t msglen = strlen(msg);
+	char * buf;
+	ssize_t r;
+
+	/* Initialize output message buffer. */
+	if ((buf = malloc(msglen)) == NULL) {
+		warnp("malloc");
+		goto err0;
+	}
+
+	/* Create socket pairs for the input and output. */
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, in)) {
+		warnp("socketpair");
+		goto err1;
+	}
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, out)) {
+		warnp("socketpair");
+		goto err1;
+	}
+
+	/* Link the two pairs. */
+	if (pushbits(in[1], out[0], &thread)) {
+		warnp("pushbits");
+		goto err1;
+	}
+
+	/* Wait for thread to start. */
+	wait_ms(100);
+
+	/* Send the message. */
```

Comment:
Maybe worth adding a comment saying that this will not block because the message is small enough to be entirely buffered within the kernel.  If `msglen` was too large we could block forever since nobody is reading the output yet.